### PR TITLE
Add counter and llamadb integration tests

### PR DESCRIPTION
**Pull request by michaelvoronov**

### CODE REVIEW

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+      res.success()
+    }
+
+    "be able to launch VM with 4 Mb memory and inserts a lot of data" in {
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
+
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath))
+        // allocate 1 МБ memory
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: MB :)

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,168 @@
+/*
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: I believe this is a wrong location for a source code file. If you want to exclude it from tests, you can `ignore` it.

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+    "be able to launch VM with 4 Mb memory and inserts a lot of data" in {
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
+
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath))
+        // allocate 1 МБ memory
+        _ <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
+        result <- vm.invoke[IO](None, "do_query", ("insert into USERS values('A')" + ",('1')"*1024*10).getBytes())
+        state <- vm.getVmState[IO].toVmError
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: Check `state` to be correct or change it to `_` if you don't use it.

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+import cats.data.EitherT
+import cats.effect.IO
+import fluence.crypto.{Crypto, CryptoError, DumbCrypto}
+import fluence.vm.TestUtils._
+import fluence.vm.VmError._
+import org.scalatest.{Matchers, WordSpec}
+
+import scala.language.{higherKinds, implicitConversions}
+
+class LlamadbIntegrationTest extends WordSpec with Matchers {
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: use trait `EitherValues`

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
+
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath))
+        // allocate 1 МБ memory
+        _ <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
+        result <- vm.invoke[IO](None, "do_query", ("insert into USERS values('A')" + ",('1')"*1024*10).getBytes())
+        state <- vm.getVmState[IO].toVmError
+      } yield {
+        result should not be None
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: `shouldBe defined`

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+        val resultAsString = new String(result.get)
+        resultAsString startsWith "rows inserted"
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: `should startWith "rows inserted"`

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      }
+
+      res.success()
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: Better use `EitherValues` like this
```
    "be able to operate with an empty strings" in {
      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath

      (for {
        vm <- WasmVm[IO](Seq(llamadbFilePath))
        _ <- vm.invoke[IO](None, "do_query", "".getBytes())
        result <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
        _ <- vm.invoke[IO](None, "do_query", "".getBytes())
        _ <- vm.getVmState[IO].toVmError
      } yield {
        result shouldBe defined

        val resultAsString = new String(result.get)
        resultAsString should startWith "rows inserted"

      }).value.unsafeRunSync().right.value
    }
```

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    "be able to launch VM with 100 Mb memory and inserts a lot of data" in {
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
+
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath), "fluence.vm.client.100Mb")
+        _ <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
+
+        // this config provides at 25 times more memory as the fluence.vm.client
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: Not sure what this comment means. Should it be something like `use 25 times more memory than previous test`?

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+import org.scalatest.{Matchers, WordSpec}
+
+import scala.language.{higherKinds, implicitConversions}
+
+class LlamadbIntegrationTest extends WordSpec with Matchers {
+
+  "llamadb integration tests" should {
+
+    "be able to operate with an empty strings" in {
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: You seem to calculate this path every time. Maybe move it to class level and reuse in each test?

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        vm <- WasmVm[IO](Seq(llamadbFilePath), "fluence.vm.client.100Mb")
+        _ <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
+
+        // this config provides at 25 times more memory as the fluence.vm.client
+        result <- vm.invoke[IO](None, "do_query", ("insert into USERS values('A')" + ",('1')"*1024*10*25).getBytes())
+        state <- vm.getVmState[IO].toVmError
+      } yield {
+        result should not be None
+
+        val resultAsString = new String(result.get)
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: I would refactor this repetitive piece to some function, e.g.,
```
def checkResult(result: Array[Byte]): Unit = {
    val resultAsString = new String(result.value)
    resultAsString should startWith "rows inserted"
}
```

File: vm/src/it/scala/LlamadbIntegrationTest.scala (hunk 4 above)
Comment (folex) [reply vm/src/it/scala/LlamadbIntegrationTest.scala]: also, use `OptionValues`

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath))
+        _ <- vm.invoke[IO](None, "do_query", "".getBytes())
+        result <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
+        _ <- vm.invoke[IO](None, "do_query", "".getBytes())
+        state <- vm.getVmState[IO].toVmError
+      } yield {
+        result should not be None
+
+        val resultAsString = new String(result.get)
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: Don't use `Option::get`, there is `.value` in `OptionValues` trait, it has better error reporting for testing purposes.

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+      res.success()
+    }
+
+    "be able to launch VM with 2 Gb memory and inserts a lot of data" in {
+      val llamadbFilePath = getClass.getResource("/src/it/resources/llamadb/llama_db_version_0_1.wasm").getPath
+
+      val res = for {
+        vm <- WasmVm[IO](Seq(llamadbFilePath), "fluence.vm.client.2Gb")
+        _ <- vm.invoke[IO](None, "do_query", "create table USERS(name varchar(1))".getBytes())
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: I would move table creation to a separate method, and reuse it in tests. DRY and less boilerplate.

File: vm/src/it/scala/LlamadbIntegrationTest.scala [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+        // this config provides at 19 times more memory as the fluence.vm.client.100Mb
+        result <- vm.invoke[IO](None, "do_query", ("insert into USERS values('A')" + ",('1')"*1024*10*25*19).getBytes())
```
Comment (folex) [inline vm/src/it/scala/LlamadbIntegrationTest.scala]: Also, I would refactor insertion to a separate function, e.g.,
```
    def insert(vm: WasmVm, megabytes: Int) = {
      val times = 1024 * megabytes
      vm.invoke[IO](None, "do_query",  ("insert into USERS values('A')" + ",('1')" * times).getBytes())
    }
```
code would become a lot cleaner and more readable, IMO.

File: vm/src/main/resources/reference.conf [hunk 14]
```diff
@@ -5,7 +5,7 @@
 fluence.vm.client {
 
   # The maximum number of memory pages when a module doesn't say
-  # Each page contains 65536 bytes of data, `65536*16=1MB
+  # Each page contains 65536 bytes of data, `65536*64=4MB
```
Comment (folex) [inline vm/src/main/resources/reference.conf]: Why that has changed? Maybe reflect the change in comments?

File: vm/src/test/scala/fluence/vm/AsmleWasmVmSpec.scala [hunk 15]
```diff
@@ -57,7 +57,7 @@ class AsmleWasmVmSpec extends WordSpec with Matchers {
```
Comment (folex) [inline vm/src/test/scala/fluence/vm/AsmleWasmVmSpec.scala]: Typo in `Asmle`. Should be `AsmbleWasmVmSpec`.

File: vm/src/main/resources/reference.conf (hunk 14 above)
Comment (michaelvoronov) [reply vm/src/main/resources/reference.conf]: Actually, the comment has been changed :). The previous one reflects the old setting and hasn't been changed when the initial one changed.

**Pull request merged**